fix pod-to-pod MTU drop when both in+egress proxy and IPSec enabled

smagnani96 opened the pull request:
This PR fixes the pod-to-pod traffic being dropped because of using a higher MTU value. This is caused by a non-configuration of the routes from proxy, in the routing table n. 2005 respectively. While the pod-to-pod route is being adjusted according to the IPSec overhead and the adjusted size of the authentication key, the from-proxy route is not changed as well.

This also enables the `pod-to-pod-with-l7-policy-encryption` test for IPSec in IPv4. Such test is therefore skipped if/for:
1.  IPv6, please refer to the follow-up issue #35485;
2. cilium version < 17.0.0, need to backport the fix before enabling the test.

Fixes: #33168

```release-note
Fix packet drops for pod-to-pod connections that pass through ingress & egress proxy when using IPsec, caused by MTU misconfiguration.
```

smagnani96 committed:
fix pod-to-pod MTU drop when both in+egress proxy and IPSec enabled

This commit fixes the pod-to-pod traffic being dropped because of using a
higher MTU value. This is caused by a non-configuration of the default route
from proxy, in the routing table n. 2005 respectively. While the pod-to-pod
default route MTU is being adjusted according to the IPSec overhead and the
adjusted size of the authentication key, the from-proxy route is not changed
as well leading to connectivity issues when both ingress and egress policies
are in-place.

Fixes: #33168

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

julianwiedmann commented on cilium-cli/connectivity/builder/pod_to_pod_encryption.go:
```diff
@@ -25,17 +26,17 @@ func (t podToPodEncryption) build(ct *check.ConnectivityTest, _ map[string]strin
+			WithScenarios(
+				tests.PodToPodEncryption(features.RequireEnabled(features.EncryptionPod)),
+			)
+	}
```
drive-by - we'll need to guard the IPsec part here to only run against v1.17 releases, until we've backported the fix into stable branches. Otherwise the test will for example run against an unfixed `v1.14` branch, and fail.

It probably also makes sense to split this change into two patches (Cilium agent vs CLI), so that it's easier to drop the CLI change when backporting.

smagnani96 replied:
You're right, thanks for the catch!

jschwinger233 commented on pkg/proxy/proxy.go:
```diff
@@ -528,16 +528,21 @@ func (p *Proxy) SetProxyPort(name string, proxyType types.ProxyType, port uint16
+	// Use RouteMTU only when both ingress and egress proxy.
+	if !fromIngressProxy || !fromEgressProxy {
+		mtu = 0
+	}
```
This reminds me: the reason in some situation we don't install 2005 rules (`!fromXgressProxy`), is tunnel routing mode already has `default dev cilium_host` to route proxy traffic. In this case we need to make sure those default routes rather than 2005 rules have correct MTU.

But it could be a bad idea to affect all the other normal from-host traffic, I need to think about this.

What did CI say? Any failure in tunnel scenarios for pod-to-pod-with-l7-policy-encryption?

smagnani96 replied:
To be honest, I was not considering the overlay scenario when working on this patch. Trying to play locally with Overlay doesn't seem to report any error to connectivity even before this patch (ESP packet not fragmented). Though, I would like to replicate locally with the exact kernel that are failing and see what happens with this test enabled.

AFAIU, CI seems to confirm this, connectivity tests are passing. The plain text IPv6 packet we're recording should be an exception and should not indicate that the whole communication was plain text (right?): I would expect more than one entry in the bpftool output otherwise, let me know if I'm mistaken.

jschwinger233 replied:
> The plain text IPv6 packet we're recording should be an exception and should not indicate that the whole communication was plain text.

Yes. It's a transient issue coming from...maybe from pod deletion, policy creating and deleting....

Can you open another draft PR to run the pod2pod encryption test, but without any other changes you made here such as routing MTU. I suspect the leak isn't caused by your PR, it's been there since forever.

jschwinger233 replied:
Withdraw what I said above.... without your changes pod2pod encryption test can't pass connectivity checks....

smagnani96 replied:
> Can you open another draft PR to run the pod2pod encryption test, but without any other changes you made here such as routing MTU. I suspect the leak isn't caused by your PR, it's been there since forever.

Yep sure, I wanted to test CI local to avoid opening a draft PR :joy: I'll open it in a minute. I suspect too it is not affected by our MTU patch, but still addressing it in this issue would be a good option.

smagnani96 replied:
> Withdraw what I said above.... without your changes pod2pod encryption test can't pass connectivity checks....

Not in native routive, but it still should pass in geneve/vxlan (apparently not in vxlan), right?

smagnani96 replied:
https://github.com/cilium/cilium/actions/runs/11273423172

1. `rhel8-20240404.144247` failing with same reason
2. `5.15-20240710.064909` failing but immediately after the 1st connectivity test (no after rotation)
3.  `bpf-next-20240711.013133` passing :open_mouth: 

All the other ones are failing because there's no the commit with the MTU fix, I was focusing on these 3.

smagnani96 replied:
Resolving for now, moving the whole conversation concerning the tests in a new issue+PR.

jschwinger233 reviewed (commented):
Re https://github.com/cilium/cilium/actions/runs/11240829491/job/31251421453:

I found the leak packets are all TCP reset:

```
 Error: bpftrace output is not empty
[18:14:44:129639] fd00:10:244:1::f085:34820 -> fd00:10:244:3::82cc:8080 (proto: 6, TCP flags: ...R, encap: 1, ifindex: 10, netns: f0000098, override: 0)
```

(Love this tcp flag information, love Marco!)

So it's probably caused by ipcache race issue when pods are removed but packets are still lingering around, maybe there is a small window when ipcache entry for that lingered packet is gone. We saw the similar issue before but forgot how we worked around it...

smagnani96 committed:
add tests for from-proxy routes to match lower MTU

This commit extends the routes tests to check that, according to the previous
commit, the default route in the table 2005 is correctly installed with the
provided MTU. This is particularly useful as when enabling both ingress
and egress policies with IPSec, the default route has a lower MTU since
it needs to account also for the encryption overhead.
With this commit, in the tests we make sure that, when provided, the correct
value is used rather than the default one.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
enabling IPv4 pod-to-pod-with-l7-policy-encryption for IPSec

This commit enables the pod-to-pod-with-l7-policy-encryption for IPSec in
IPv4 only. With the previous commits, we fix the MTU when having both
ingress and egress proxy and policies, re-establishing the connectivity
between pods. This test checks that this holds true for IPv4.
Unfortunately, the IPv6 test seems to be flaky, causing a plain TCP RST packet
to be sent without encryption when running with vxlan. This behavior is
investigated in a separate issue (see #35485).

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

jschwinger233 commented on cilium-cli/connectivity/tests/encryption.go:
```diff
@@ -197,6 +197,26 @@ func isWgEncap(t *check.Test) bool {
+	}
+	if !versioncheck.MustCompile(">=1.17.0")(t.Context().CiliumVersion) {
+		return fmt.Errorf("enabling test for IPSec requires backporting")
+	}
```
feature and version checks can be put in cilium-cli/connectivity/builder/pod_to_pod_encryption.go to skip the whole test at all with clear skip reason. Like what you did at https://github.com/cilium/cilium/pull/35486/files#diff-7a17abdbd1cb8e727bf901ec2001452944a48e4d03d83a46ee7e8e760fe9e6b3R30-R39

smagnani96 replied:
Ok, but then I don't think the following is enough [here](https://github.com/cilium/cilium/blob/main/cilium-cli/connectivity/tests/encryption.go#L347), as it would prevent also WireGuard+IPv6:

```go
if ipFam == features.IPFamilyV6 {
    // comment, #issue_no
    return
}
```

I would need to add:

```go
if f, ok := t.Context().Feature(features.EncryptionPod); ok && f.Enabled && f.Mode == "ipsec" && ipFam == features.IPFamilyV6 {
    // comment, #issue_no
    return
}
		
```

and that is why I decided to create a unique function where to check both the requirements (cilium version and ipFamily with IPSec).

jschwinger233 replied:
You are correct. Resolve the conversation.

smagnani96 replied:
Let me know what is the preferred way, I'm ok with both. If we want to split the checks between the two files (version in `pod_to_pod_encryption.go` and IPSec+IPv6 in `encryption.go`) I'll correct the last commit :+1:

jschwinger233 reviewed (approved):
Thank you! Keep an eye on https://github.com/cilium/cilium/actions/runs/11477726748/job/31940507230 (and ci-e2e for wireguard), check the log and make sure `pod-to-pod-with-l7-policy-encryption` does get enabled.
